### PR TITLE
Add 2fa token text field to login

### PR DESCRIPTION
**Pull request by justincmoy**
# Checklist
- [x] I have described what this PR contains

**Choose one of the following two options:**
- - [x] This PR does not introduce major changes
- - [ ] This PR introduces major changes, and I have consulted @buresdv, @jboi or @mormaer in the *Mlem Development Matrix room*

**Choose one of the following two options:**
- - [ ] This PR does not change the UI in any way
- - [x] This PR adds new UI elements / changes the UI, and I have attached pictures or videos of the new / changed UI elements

# Pull Request Information

## About this Pull Request
**Describe what this PR contains in as much detail as possible**

Following the (eventual [0],[1],[2]) addition of a totp token to the login workflow, this adds a text field upon receiving a specific error message from the server.

[0] https://github.com/LemmyNet/lemmy/pull/2741
[1] https://github.com/LemmyNet/lemmy-js-client/pull/106
[2] https://github.com/LemmyNet/lemmy-ui/commit/af1e8868bd13df441ea02f9c13f5c3ee359db597

## Screenshots and Videos
**In case this PR changes something in the UI, please include screenshots or videos of this new feature**

The following video includes:
- Showing and focusing the 2FA token text field following login that requires the token, without showing the error returned by the server.
- Showing the error message returned by the server if the 2FA token is incorrect.
- Successfully logging in as a user with a correct 2FA token.

https://github.com/buresdv/Mlem/assets/14875982/60b9c9bb-2ac3-4382-938a-92693075f164

## Additional Context
**Any additional context you'd like to add to help us review this PR**

### CODE REVIEW

**Comment by mormaer** [comment]
Nice, we should definitely get this supported and should be simple enough to integrate with the changes in #95 👍 

There is some basic error handling in the other PR which would catch that error, but currently it just wraps the error message from the API (`"missing_totp_token"`) and returns it for handling... I agree that it should likely return it's own bespoke error for this case.

It's unfortunate that the error model returned is just a `String`, as determining behaviour using a string comparison is less than ideal, but the API is the API and for now that may be the best option we have 😞

**Comment by buresdv** [comment]
This is a great addition. I believe that the 2fa field should be hidden by default and only show up if the server responds with the "missing token" message. That way, users that don’t have it configured don’t have to leave an empty field. Users expect to have to fill in every available field.

**Comment by justincmoy** [comment]
> It's unfortunate that the error model returned is just a `String`, as determining behaviour using a string comparison is less than ideal, but the API is the API and for now that may be the best option we have 😞

I agree. If anything, this PR could do a better job as the `missing_totp_token` String is a value of the `.error` JSON key, so perhaps we could attempt to unmarshal the response body to json, and then switch-case on the error key.

> I believe that the 2fa field should be hidden by default and only show up if the server responds with the "missing token" message.

That's how this PR currently behaves via the `isShowingTwoFactorSection` flag.

**Comment by buresdv** [comment]
> That's how this PR currently behaves via the `isShowingTwoFactorSection` flag.

Great, I didn't notice that. Maybe one more suggestion would be to not show the "Please enter the 2FA token" as an error (because some users might get confused by us showing an error while it's not really an error, just a request for more information) and instead just reveal and focus the 2FA field?

**Comment by ShadowJonathan** [comment]
What if you show the "please enter the 2FA token" as a popup with an entry field?

**Comment by justincmoy** [comment]
> > That's how this PR currently behaves via the `isShowingTwoFactorSection` flag.
> 
> Great, I didn't notice that. Maybe one more suggestion would be to not show the "Please enter the 2FA token" as an error (because some users might get confused by us showing an error while it's not really an error, just a request for more information) and instead just reveal and focus the 2FA field?

Done!

File: Mlem/Views/Tabs/Posts/Components/Post/Components/Instance and Community List/Add Account View.swift [hunk 1]
```diff
@@ -137,6 +143,24 @@ struct AddSavedInstanceView: View
...
+                    if isShowingTwoFactorText {
+                        HStack
+                        {
+                            Text("2FA Token")
+                            Spacer()
+                            SecureField("TwoFactorToken", text: $twoFactorToken, prompt: Text("000000"))
+                                .focused($focusedField, equals: .twoFactorField)
+                                .submitLabel(.go)
+                                .keyboardType(.asciiCapableNumberPad)
+                                .onAppear
```
Comment (WestonHanners) [inline Mlem/Views/Tabs/Posts/Components/Post/Components/Instance and Community List/Add Account View.swift:156]: `.textContentType(.oneTimeCode)`
https://developer.apple.com/documentation/uikit/uitextcontenttype/2980930-onetimecode

**Review by mormaer — APPROVED** [review]
All good from me, the `+4` second delay hurts my soul but until we can refactor this view it's a necessity - happy for this to be merged if you are @WestonHanners, we could always add the accessibility modifier for one time code as a follow-up?

**Pull request merged**